### PR TITLE
Add state locking

### LINKED ISSUE #448: The animation to teleport back to the Ship using the post-game map menu SHIP option can be interrupted
Since the animation to teleport back to the Ship uses the state machine, anything that sets the gamestate can interrupt it. This includes falling onto a teleporter, as that sets the gamestate to 2000, or even falling into another room, as in non-finalmode and non-custommode the gamestate is set to 0 whenever you do a screen transition (if you've never done the Comms Relay cutscene).

Potentially, there are also script boxes that the player could move into after activating the animation and interrupt the animation that way, too, but this seems hard to pull off without also triggering a screen transition that would've interrupted the animation anyways.

You can also temporarily interrupt the animation by starting it right before falling into or coming into contact with a hazard. It'll only be a temporary interruption if you respawn in the same room, at which point the animation will resume right after the death animation is done. But it'll be a complete interruption if you respawn in a different room, because the gotoroom will set the gamestate to 0, as explained above.

### PR DESCRIPTION
**Pull request by AllyTally**
## Changes:

This PR adds a few new functions to manage changing the gamestate (`game.state`). While you're teleporting to the ship, it prevents the game from setting a new state until the initial teleport animation is finished. This is done using `lockstate()`, `unlockstate()`, and the generic teleport animation setting the state _manually_.

Fixes #448.

## Legal Stuff:

By submitting this pull request, I confirm that...

- [x] My changes may be used in a future commercial release of VVVVVV
- [x] I will be credited in a `CONTRIBUTORS` file and the "GitHub Friends"
  section of the credits for all of said releases, but will NOT be compensated
  for these changes

### CODE REVIEW

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:373]: Not sure why this blank line was added.

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:720]:
```suggestion
void Game::setstate(const int gamestate)
```

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:728]:
```suggestion
void Game::setstate(const int gamestate, const int delay)
```

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:745]:
```suggestion
void Game::setstatedelay(const int delay)
```

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:766]:
```suggestion
    if (statedelay <= 0)
    {
```

File: desktop_version/src/Script.cpp
Comment (InfoTeddy) [inline desktop_version/src/Script.cpp:1120]:
```suggestion
                // Allow the gamestate command to bypass statelock, at least for now
```

**Review by InfoTeddy — CHANGES_REQUESTED** [review]
Style pass.

I also noticed that `game.state = activetrigger;` wasn't changed to use the `setstate` function, it should probably do that or otherwise the teleport animation can be interrupted by the player moving through a script box.

**Review by InfoTeddy — CHANGES_REQUESTED** [review]
I forgot to check for `statedelay` assignments last time (my bad), so here are the gamestates of `statedelay` assignments that haven't been changed to use `setstatedelay` and should be:

* 3502
* 3504
* 3506
* 3507

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:2946]: I don't think this ends up causing any issues, but why was this changed to `state++;`? It should be left the same.

**Review by InfoTeddy — CHANGES_REQUESTED** [review]
Another style pass.

I also noticed that the overload of `setstate` that sets both `state` and `statedelay` isn't used outside of `Game.cpp`. I presume this was done for clarity so it'd be easy to tell that the state delay is being set alongside the state as well. But if that's the case, then why not just axe the overload and go back to calling both functions separately?

**Comment by AllyTally** [comment]
Whoops, the `state++;` was an accident

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:339]: Overzealous find-and-replace? This should stay as `statedelay = 0;` because it's an init function.

**Review by InfoTeddy — CHANGES_REQUESTED** [review]
One more style nitpick before I merge this.

**Comment by AllyTally** [comment]
Should be good now

File: desktop_version/src/Game.cpp
Comment (InfoTeddy) [inline desktop_version/src/Game.cpp:339]: No, this is incorrect. `setstatedelay` is a function, not a variable.
```suggestion
    statedelay = 0;
```

**Review by InfoTeddy — CHANGES_REQUESTED** [review]

File: desktop_version/src/Game.cpp
Comment (AllyTally) [reply desktop_version/src/Game.cpp]: agh, sorry, was distracted while committing

Comment (InfoTeddy) [reply desktop_version/src/Game.cpp]: lmao

**Review by InfoTeddy — APPROVED** [review]
Will merge once CI passes.

**Pull request merged**
